plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/CallSiteInline.hs

```diff
@@ -164,25 +104,28 @@ fullyApplyAndBetaReduce info args0 = do
         name ->
         -- `arg`
         Term tyname name uni fun ann ->
+        -- the body `a` will be beta reduced in
+        Term tyname name uni fun ann ->
         InlineM tyname name uni fun ann Bool
-      safeToBetaReduce a arg = shouldUnconditionallyInline Strict a arg rhsBody
-  go rhsBody (varArity info) args0
+      safeToBetaReduce = shouldUnconditionallyInline Strict
+  go rhs args0
 
 -- | Consider whether to inline an application.
-inlineSaturatedApp ::
+inlineApp ::
   forall tyname name uni fun ann.
   (InliningConstraints tyname name uni fun) =>
   Term tyname name uni fun ann ->
   InlineM tyname name uni fun ann (Term tyname name uni fun ann)
-inlineSaturatedApp t
+inlineApp t
   | (Var _ann name, args) <- splitApplication t =
       gets (lookupVarInfo name) >>= \case
-        Just varInfo ->
-          fullyApplyAndBetaReduce varInfo args >>= \case
-            Just fullyApplied -> do
-              let rhs = varRhs varInfo
-                  -- Inline only if the size is no bigger than not inlining.
-                  sizeIsOk = termSize fullyApplied <= termSize t
+        Just varInfo -> do
+          -- rename the rhs of the variable before any substitution
+          rhs <- liftDupable (let Done rhs = varRhs varInfo in rhs)
+          applyAndBetaReduce rhs args >>= \case
+            Just applied -> do -- there is beta reduction of the application
+              let -- Inline only if the size is no bigger than not inlining.
+                  sizeIsOk = termSize applied <= termSize t
                   -- The definition itself will be inlined, so we need to check that the cost
```

Comment:
This logic is no longer correct now that we do beta reduction also. It's not totally clear to me what the correct logic for the cost actually is now :thinking:

Reply:
@zliu41 any thought on this? I think maybe we don't need to check the cost at all? We already made sure it's safe to beta reduce - ie we made sure no effects etc.

Reply:
It is still needed, and I think it should still be correct.

> We already made sure it's safe to beta reduce 

That's different. Consider

```
let f = 1+2+3+...+1000
 in ...f...f...
```

Here `f` is not even a function, so `safeToBetaReduce` does not apply. But we can't inline `f`, because it will duplicate the costly computation (in this case, the size of `f` is also too big, but it is possible for size to be small but cost to be big).

Reply:
I believe currently we don't inline non-functions in the call site inliner. See [here](https://github.com/input-output-hk/plutus/blob/99d4253dc071c68aa2dcaa3746d2feae2118a10f/plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/CallSiteInline.hs#L98). So that is not a relevant case? Any function (with term arguments) would go through `safeToBetaReduce`.

Reply:
right but there's no actual substitution in this case.

Reply:
The question here is whether checking the cost is needed. There is no substitution, but I think it is still needed, because of my example.

Reply:
Here we are talking about the check for the call site inliner. If you look at the line I [linked](https://github.com/input-output-hk/plutus/blob/99d4253dc071c68aa2dcaa3746d2feae2118a10f/plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/CallSiteInline.hs#L98), no arguments are substituted, if the variable is a non-function to begin with. That was what I meant.

I can try with a test yes but I won't go up to 1000 lol.

Reply:
But yeah I see it will still look at that because it's returned as `applied`.

Reply:
but looking at that maybe wrong in the function case if it has been beta reduced, like Michael said. So maybe we should deal with non-function separately afterall?

Reply:
of course, we could just keep the check as is and inline less.

Reply:
maybe we can check `costIsAcceptable applied` instead?

Reply:
tested in https://github.com/input-output-hk/plutus/pull/5426/commits/127b9d382c3dd7d4ecfd8135c308f25bfcad48d3 not much difference.

Reply:
actually there is regression when I apply the change https://github.com/input-output-hk/plutus/pull/5426/commits/11d27b41f6a05f3729019603e8023c11c1cc89a9 :( I reverted it for now.